### DISCUSSION TITLE
Add sound

### DISCUSSION THREAD

**Discussion by udo-munk** [udo-munk/z80pack#318] [Ideas]
There is an open issue, that the z80pack machines won't output sound, when you toggle switches, the head of a disk drive moves, and so on. This is not really an issue, this is a feature that could be added, so I move this here to discuss and close the issue.

The machines run on a wide range of POSIX compatible workstations. The last time I looked at sound support for OSX, Linux, Windows there was no common standard, not even something to create simple tones. So this might be quite challenging, would be OK to first implement it on one OS of course.

**Comment by Selorax** [comment]
I think it would be a nice feature to have. I don't know what to do with a problem of absence of common sound API, but somehow the problem of interface, requiring X11 on GNU/Linux, Cocoa on Mac and Windows32api on Windows was solved, so I guess it should be doable. 
Linux used to support OSS, I don't know if it was dropped completely or just disabled by default in most distribution, but apparently modern linux does not have /dev/dsp anymore.

**Comment by udo-munk** [comment]
Sure, could be done. Requires a team of dedicated people who know all and everything about the low level sound interface at driver level on Windows, OSX and Linux/BSD.... All different of course, each with it's own problems and workarounds. Then you define a API that is totally vendor independent and makes use of the low level interface of the platform. Because right now everyone has their own API, totally different und incompatible of course. And then you can make use of this API in applications. Can all be done, certainly not by me in this life anymore.

During the 90th I worked on adding the low level sound interface into UNIX systems. After all was working well the next generation of sound cards was thrown out, nothing was working anymore und you started all over again. So I know a little bit about what one is looking into for stuff like this ;-)

I also used to work with the OSS guys a bit, because they supported an API that was UNIX brand independend, your application using this API would run on Linux, BSD, SCO OpenServer, SCO Unixware, Solaris. Really cool stuff, but for some reason or another gone by now.

**Comment by Selorax** [comment]
As far as I know BSD retains OSS. According to this article https://alsa.opensrc.org/OSS_emulation Linux still supports OSS api via emulation, `modprobe snd-pcm-oss` was enough to create /dev/dsp and it works if I use `cat /dev/urandom > /dev/dsp`  so maybe OSS support is going to be a good starting point, since it will work for two major desktop OSes, namely GNU/Linux and BSD, with Windows and macOS remaining. 

The reason why OSS disappeared was that OSS v4 which finally supported simultaneous sound from several applications became proprietary and was released as open source only later. Same goes with Plan9 by the way.

I think that's the most likely result why those programs didn't get enough traction.

**Comment by udo-munk** [comment]
The problem with OSS was that their low level drivers were proprietary. The API to use them of course not. They gave me a free license for all drivers on any platform I wished to use, so that I was able to add sound support to applications. Search for xdoom, that is one of the applications using it.
Anyway, been there, done that, got t-shirt, so this is for anyone else who would like to get into this.

**Comment by sneakywumpus** [comment]
I started converting z80pack from using X11 to SDL2. Still work in progress ([sdl2 dev branch](https://github.com/sneakywumpus/z80pack/tree/sdl2)). This will make it possible to add platform independent sound support.

**Comment by udo-munk** [comment]
It would be better to add a version using SDL parallel and independent from the X11 version, I did that once with xdoom and both were useful. So software still runs on platforms without SDL but X11. I'm not sure if SDL2 is really thread save nowadays, with X11 it needed ages, didn't follow SDL development for a longer time anymore though, was the typical game half backed not properly working stuff.

**Comment by sneakywumpus** [comment]
I'll redo this as optional SDL support, i.e. adding `WANT_SDL=YES` to the machine Makefile generates the SDL version, otherwise the X11 version.

**Comment by udo-munk** [comment]
Excellent, then I still can use it sometimes on large IBM or Sun systems and see how it performs on these CPU's. One has the X11 runtime of course, but propably no one can be bothered with a SDL port.

**Comment by sneakywumpus** [comment]
This time SDL2 is added as an option. See ([addsdl2 dev branch](https://github.com/sneakywumpus/z80pack/tree/addsdl2)). Compile with `WANT_SDL=YES make` . Just `make` uses X11. Still work in progress.

**Comment by sneakywumpus** [comment]
Works well now. Check it out with SDL2 on your Mac. See docs/README-osx.txt

**Comment by sneakywumpus** [comment]
Hmm.. Something is wrong with the native OpenGL on my MacBook 12". The Imsai 3-D model renders strange. Strange stripes on the sides. Can see switches thru the top...

**Comment by udo-munk** [comment]
Tested some on WSL/Ubunto, the 3D IMSAI model also looks the same stange. But 2D models are working fine, looks good. The latest mesa update on Ubunto again broke the 3D models, as usual, so in the long run SDL propably is a good alternative. I think you can create PR for dev upstream if you want.

**Reply by udo-munk** [reply]
On MacOS 14.6.1 the IMSAI 3D model looks OK, no odd drawing of the box.

**Comment by udo-munk** [comment]
Trying on MacOS but I'm missing something

udo@MacBook-Pro-von-Udo z80pack % cd altairsim 
udo@MacBook-Pro-von-Udo altairsim % ./cpm22
dyld[70341]: Library not loaded: @rpath/SDL2_image.framework/Versions/A/SDL2_image
  Referenced from: <5FC2B3CC-DF34-3A2E-9807-77696CFAC359> /Users/udo/work/TEST/z80pack/altairsim/altairsim
  Reason: no LC_RPATH's found
./cpm22: line 8: 70341 Abort trap: 6           ./altairsim $*

**Comment by udo-munk** [comment]
So after some reading looks like one needs to do:

install_name_tool -add_rpath /Library/Frameworks ./altairsim 

So far so good, now the complains are:
<img width="409" alt="Bildschirmfoto 2025-01-08 um 18 31 59" src="https://github.com/user-attachments/assets/d925a776-1c62-4fb1-88a7-944b0288055e" />

**Comment by sneakywumpus** [comment]
I've haven't had any problems on my MacBook 12-inch Early 2016 with macOS 12.7.6 and Xcode 14.2 (I don't have access to any newer Mac), no messages whatsoever.... I'll look around on the web for macOS SDL2 projects and see how they do it.

Under "System Preferences / Security & Privacy / Privacy" under "Developer Tools" I have "Terminal" enabled as allowed to run software that does not meet the system's security policy,  and also I have "Terminal" enabled under "Full Disk Access".

**Reply by udo-munk** [reply]
Yep, one has to change the security policy for the terminal application, then it works.

**Comment by sneakywumpus** [comment]
I have added `-Wl,-rpath,/Library/Frameworks`  to the Makefiles for macOS.

**Reply by udo-munk** [reply]
This helped, finds the libs now.

**Comment by sneakywumpus** [comment]
I changed the OpenGL depth buffer from the SDL2 default of 16-bit to 24-bit, that made it work correctly.
